src/interrupt/mod.rs

```diff
@@ -0,0 +1,217 @@
+//! MSI interrupts.
+//!
+//! A device may support multiple types of interrupts, and each type of interrupt may support one
+//! or multiple interrupt sources. For example, a PCI device may support:
+//! * Legacy Irq: exactly one interrupt source.
+//! * PCI MSI Irq: 1,2,4,8,16,32 interrupt sources.
+//! * PCI MSIx Irq: 2^n(n=0-11) interrupt sources.
+//!
+//! A distinct Interrupt Source Identifier (ISID) will be assigned to each interrupt source.
+//! An ID allocator will be used to allocate and free Interrupt Source Identifiers for devices.
+//! To decouple the vm-device crate from the ID allocator, the vm-device crate doesn't take the
+//! responsibility to allocate/free Interrupt Source IDs but only makes use of assigned IDs.
+//!
+//! The overall flow to deal with interrupts is:
+//! * the VMM creates an interrupt manager
+//! * the VMM creates a device manager, passing on an reference to the interrupt manager
+//! * the device manager passes on an reference to the interrupt manager to all registered devices
+//! * guest kernel loads drivers for virtual devices
+//! * guest device driver determines the type and number of interrupts needed, and update the
+//!   device configuration
+//! * the virtual device backend requests the interrupt manager to create an interrupt group
+//!   according to guest configuration information
+
+use std::sync::Arc;
+use vmm_sys_util::eventfd::EventFd;
+
+/// Reuse std::io::Result to simplify inter operability among crates.
+pub type Result<T> = std::io::Result<T>;
+
+/// Data type to store an interrupt source identifier.
+pub type InterruptIndex = u32;
+
+/// Maximum number of global interrupt sources.
+pub const MAX_IRQS: InterruptIndex = 1024;
+
+#[cfg(feature = "generic_msi")]
+/// Maximum number of Message Signaled Interrupts per device.
+pub const MAX_MSI_IRQS_PER_DEVICE: InterruptIndex = 32;
+
+/// Type of interrupt source.
+pub enum InterruptSourceType {
+    #[cfg(feature = "legacy_irq")]
+    /// Legacy Pin-based Interrupt.
+    /// On x86 platforms, legacy interrupts are routed through 82599 PICs and/or IOAPICs.
+    LegacyIrq,
+    #[cfg(feature = "msi_irq")]
+    /// Message Signaled Interrupt (PCI MSI/PCI MSIx/Generic MSI).
+    /// Some non-PCI devices (like HPET on x86) make use of generic MSI in platform specific ways.
+    MsiIrq,
+}
+
+/// Configuration data for an interrupt source.
+pub enum InterruptSourceConfig {
+    #[cfg(feature = "legacy_irq")]
+    /// Configuration data for Legacy interrupts.
+    LegacyIrq(LegacyIrqSourceConfig),
+    #[cfg(feature = "generic_msi")]
+    /// Configuration data for GenericMsi, PciMsi, PciMsix interrupts.
+    MsiIrq(MsiIrqSourceConfig),
+}
+
+#[cfg(feature = "legacy_irq")]
+/// Configuration data for legacy interrupts.
+///
+/// On x86 platforms, legacy interrupts means those interrupts routed through PICs or IOAPICs.
+pub struct LegacyIrqSourceConfig {}
```

Comment:
If this is empty, then we even don't need it.

Reply:
This helps to make code it symmetric:)